src/librustc_trans/trans/context.rs

```diff
@@ -977,11 +942,62 @@ fn declare_intrinsic(ccx: &CrateContext, key: &str) -> Option<ValueRef> {
                 ccx.intrinsics().borrow_mut().insert($name, f.clone());
                 return Some(f);
             }
+        );
+        ($name:expr, $cname:ident ($($arg:expr),*) -> $ret:expr) => (
+            if !is_like_pnacl {
+                // The `if key == $name` is already in ifn!
+                ifn!($name, fn($($arg),*) -> $ret);
+            } else if key == $name {
+                let f = declare::declare_cfn(ccx, stringify!($cname),
+                                             Type::func(&[$($arg),*], &$ret),
+                                             ccx.tcx().mk_nil());
+                ccx.intrinsics().borrow_mut().insert($name, f.clone());
+                return Some(f);
+            }
         )
     }
 
     compatible_ifn!("llvm.assume", noop(llvmcompat_assume(i1) -> void), 6);
 
+    // These must be redirected to `libm` here so PNaCl's linker will link in
+    // these functions from `libm`.
+    compatible_ifn!("llvm.copysign.f32", copysignf(t_f32, t_f32) -> t_f32);
```

Comment:
> Except Rust is unique by the way Rust (fairly implicitly) references the builtins. Clang doesn't generate references to them unless one of the __builtin_ functions are used, and nobody does that (evidenced by the fact that this case isn't handled at all by pnacl-clang) (or at least no common OSS does it).

I'm not sure what is meant here by `fairly implicitly`, but I don't think there is anything special about Rust here? I've seen LLVM outputs from compilers like Julia, and such non-clang compilers often use an LLVM math intrinsic. They declare it and use it, and they expect LLVM to make linking and so forth just work. In other words, I believe that code sample with `__builtin_powif` is valid C code, and the LLVM IR derived from it is valid IR. If PNaCl wants to support as many languages as possible, it should support such code.

If I understood what PNaCl does here, it first links all the bitcode, then it does lowering of intrinsics to the PNaCl final wire format. And some of those lowerings might require libm, which might not have been linked in before? It sounds like this is just a limitation in PNaCl - it should do something before that linking, to determine if it will need libm later, something like option 1 that you mention. PNaCl linking is just different enough from typical native linking that something like this is necessary, since natively, libm would be linked in after compiling to object/executable.

(Emscripten btw has a similar issue of first linking all the bitcode, then doing compilation, and it also has a mechanism to handle such dependencies, that we need to be aware of early because they would otherwise only be discovered too late. We have them explicitly listed in a `deps.json` file. We actually don't have this issue with math functions, because of how we lower them, but we do have it with others.)